Added Simulated Annealing (#1371)
cp-algorithms/cp-algorithms: src/num_methods/simulated_annealing.md

---
tags:
    - Original
---

# Simulated Annealing

**Simulated Annealing (SA)** is a randomized algorithm, which approximates the global optimum of a function. It's called a randomized algorithm, because it employs a certain amount of randomness in its search and thus its output can vary for the same input.

## The problem

> We are given a function $E(s)$, which calculates the potential of the state $s$. We are tasked with finding the state $s_{best}$ at which $E(s)$ is minimized. **SA** is suited for problems where the states are discrete and $E(s)$ has multiple local minima. We'll take the example of the [Travelling Salesman Problem (TSP)](https://en.wikipedia.org/wiki/Travelling_salesman_problem). 

Reviewer: > which calculates the potential of the state $s$

Please define "potential" before using the term.

### Travelling Salesman Problem (TSP)

You are given a set of nodes in 2 dimensional space. Each node is characterised by its $x$ and $y$ coordinates. Your task is to find the an ordering of the nodes, which will minimise the distance to be travelled when visiting these nodes in that order.

### State

> State space is the collection of all possible values that can be taken by the independent variables.

Reviewer: > independent variables

Independent from what and in which sense? This also uses the term without defining it. Also wouldn't it be easier to say that the state space is the domain of the energy function, and a state is any its element?

Author: Yup went ahead with your suggestion for this.

A state is a unique point in the state space of the problem. In the case of TSP, all possible paths that we can take to visit all the nodes is the state space, and any single one of these paths can be considered as a state.

### Neighbouring state

It is a state in the state space which is close to the previous state. This usually means that we can obtain the neighbouring state from the original state using a simple transform. In the case of the Travelling Salesman Problem, a neighbouring state is obtained by randomly choosing 2 nodes, and swapping their positions in the current state. 

> ### The energy function E(s)

Reviewer: Again, the term energy is used without a definition or any kind of informal explanation as to why this specific term is applied to $E(s)$ in particular.

Author: Its analogous to the internal energy of the material undergoing annealing, which is explained in the Motivation section.

$E(s)$ is the function which needs to be minimised (or maximised). It maps every state to a real number. In the case of TSP, $E(s)$ returns the distance of travelling one full circle in the order of nodes in the state. 

## The approach

We start of with a random state $s$. In every step, we choose a neighbouring state $s_{next}$ of the current state $s$. If $E(s_{next}) < E(s)$, then we update $s = s_{next}$. Otherwise, we use a probability acceptance function $P(E(s),E(s_{next}),T)$ which decides whether we should move to $s_{next}$ or stay at $s$. T here is the temperature, which is initially set to a high value and decays slowly with every step. The higher the temperature, the more likely it is to move to $s_{next}$. 
At the same time we also keep a track of the best state $s_{best}$ across all iterations. Proceeding till convergence or time runs out.

### How does this work?

> This algorithm is called simulated annealing because we are simulating the process of annealing, wherein a material is heated up and allowed to cool, in order to allow the atoms inside to rearrange themselves in an arrangement with minimal internal energy, which in turn causes the material to have different properties. The state is the arrangement of atoms and the internal energy is the function being minimised. We can think of the original state of the atoms, as a local minima for its internal energy. To make the material rearrange its atoms, we need to motivate it to go across a region where its internal energy is not minimised in order to reach the global minima. This motivation is given by heating the material to a higher temperature. 

Reviewer: I would move this explanation much closer to the beginning of the article.

> Simulated annealing, literally simulates this process. We start off with some random state (material) and set a high temperature (heat it up). Now, the algorithm is ready to accept states which have a higher energy than the current state, as it is motivated by the high value of $T$. This prevents the algorithm from getting stuck inside local minimas and move towards the global minima. As time progresses, the algorithm cools down and refuses the states with higher energy and moves into the closest minima it has found.

Reviewer: ```suggestion
Simulated annealing, literally, simulates this process. We start off with some random state (material) and set a high temperature (heat it up). Now, the algorithm is ready to accept states which have a higher energy than the current state, as it is motivated by the high value of $T$. This prevents the algorithm from getting stuck inside local minimas and move towards the global minima. As time progresses, the algorithm cools down and refuses the states with higher energy and moves into the closest minima it has found.
```

<center>
<img src="https://upload.wikimedia.org/wikipedia/commons/d/d5/Hill_Climbing_with_Simulated_Annealing.gif" width="800px">
<br>
<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
<br>

> <i>This <a href="https://upload.wikimedia.org/wikipedia/commons/d/d5/Hill_Climbing_with_Simulated_Annealing.gif">gif</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.

Reviewer: ```suggestion
<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
```
Note that, in principle, CC0 doesn't require any attribution (but we can still keep it as a token of respect).

</center>

## Probability Acceptance Function(PAF)

> $P(E,E_{next},T) = 
>     \begin{cases}
>        \text{True} &\quad\text{if }  \exp{\frac{-(E_{next}-E)}{T}} \ge random(0,1) \\
>        \text{False} &\quad\text{otherwise}\\
>      \end{cases}$
> 
> This function takes in the current state, the next state and the Temperature , returning a boolean value, which tells our search whether it should move to $s_{next}$ or stay at $s$. Note that for $E_{next} < E$ , this function will always return True. 

Reviewer: ```suggestion
$P(E,E_{next},T) = 
    \begin{cases}
       \text{True} &\quad\text{if }  \mathcal{U}_{[0,1]} \le \exp(-\frac{E_{next}-E}{T}) \\
       \text{False} &\quad\text{otherwise}\\
     \end{cases}$

Here, $\mathcal{U}_{[0,1]}$ is a continuous uniform random value on $[0,1]$. This function takes in the current state, the next state and the temperature, returning a boolean value, which tells our search whether it should move to $s_{next}$ or stay at $s$. Note that for $E_{next} < E$ , this function will always return True, otherwise it can still make the move with probability $\exp(-\frac{E_{next}-E}{T})$, which corresponds to the [Gibbs measure](https://en.wikipedia.org/wiki/Gibbs_measure).
```

```cpp
bool P(double E,double E_next,double T){

>     double prob = (double)rand()/RAND_MAX; // Generate a random number between 0 and 1

Reviewer: I think this naming is misleading. The actual transition probability is $\exp(-\frac{E_{next}-E}{T})$, and this value is only used to emulate it. So, ideally we want to rewrite it using e.g. `std::bernoulli_distribution`.

    if(exp(-(E_next-E)/T) > prob) return true;
    else return false;
}
```
## Code Template

```cpp
class state{
    public:
    state(){
        // Generate the initial state
    }
    state(state& s){
        // Implement the copy constructor
    }
    state next(){
        state s_next = state(*this);

        // Modify s_next to the neighbouring state

        return s_next;
    }
    double E(){
        // Implement the cost function here
    };
};

pair<double,state> simAnneal(){
    state s = state();
    state best = state(s);
    double T = 1000; // Initial temperature

>     double u = 0.99; // decay rate

Reviewer: It seems to be the first time the "decay" is introduced. It should be mentioned earlier, as well as some explanation to why the rate of decay is exponential.

> class state{
>     public:
>     state(){
>         // Generate the initial state
>     }
>     state(state& s){
>         // Implement the copy constructor
>     }
>     state next(){
>         state s_next = state(*this);
> 
>         // Modify s_next to the neighbouring state
> 
>         return s_next;
>     }
>     double E(){
>         // Implement the cost function here
>     };
> };
> 
> pair<double,state> simAnneal(){
>     state s = state();
>     state best = state(s);
>     double T = 1000; // Initial temperature
>     double u = 0.99; // decay rate
>     double E = s.E();
>     double E_next;
>     double E_best = E;
>     while (T > 1){
>         state next = s.next();
>         E_next = next.E();
>         if(P(E,E_next,T)){
>             s = next;
>             if(E_next < E_best){
>                 best = s;
>                 E_best = E_next;
>             }
>         }
>         E = E_next;
>         T *= u;
>     }
>     return {E_best,best};
> }

Reviewer: ```suggestion
class state {
    public:
    state() {
        // Generate the initial state
    }
    state(state const& s) {
        // Implement the copy constructor
    }
    state next() {
        state s_next = state(*this);

        // Modify s_next to a random neighboring state

        return s_next;
    }
    double E() {
        // Implement the energy function here
    };
};

pair<double, state> simAnneal() {
    state s = state();
    state best = state(s);
    double T = 1000; // Initial temperature
    double u = 0.99; // decay rate
    double E = s.E();
    double E_next;
    double E_best = E;
    while (T > 1) {
        state next = s.next();
        E_next = next.E();
        if(P(E, E_next, T)) {
            s = next;
            if(E_next < E_best) {
                best = s;
                E_best = E_next;
            }
        }
        E = E_next;
        T *= u;
    }
    return {E_best, best};
}
```

Reviewer: Actually I doubt you even need an explicit copy constructor, usually the default one is fine?

Author: It might be required for some problems, but it is a nuisance to keep it there by default so I've removed it.

```
## How to use:

> Fill in the state class functions as appropriate. If you are trying to find a global maxima and not a minima, ensure that the $E()$ function returns negative of the function you are maximising and finally print out $-E_{best}$. Set the below parameters as per your need.

Reviewer: ```suggestion
Fill in the state class functions as appropriate. If you are trying to find a global maxima and not a minima, ensure that the $E()$ function returns negative of the function you are maximizing and print $-E_{best}$ in the end. Set the below parameters as per your need.
```

- $u$ : Decay. Decides the rate of cooling. A slower cooling rate (larger value of u) usually gives better results, at the cost of running for a longer time. Ensure $u < 1$. 

The number of iterations the loop will run for is given by the expression

$N =   \lceil -\log_{u}{T} \rceil$ 

Tips for choosing $T$ and $u$ : If there are many local minimas and a wide state space, set $u = 0.999$, for a slow cooling rate, which will allow the algorithm to explore more possibilities. On the other hand, if the state space is narrower, $u = 0.99$ should suffice. If you are not sure, play it safe by setting $u = 0.998$ or higher. Calculate the time complexity of a single iteration of the algorithm, and use this to approximate a value of $N$ which will prevent TLE, then use the below formula to obtain $T$.

$T = u^{-N}$

### Example implementation for TSP
```cpp

class state{
    public:
    vector<pair<int,int>> points;

    state(){
        {% raw %}points = {{0,0},{2,2},{0,2},{2,0},{0,1},{1,2},{2,1},{1,0}};{% endraw %} 
    }
    state(state& s){
        points = s.points;
    }
    state next(){
        state s_next = state(*this);

>         int a = (rand()%points.size());

Reviewer: Using `rand()` is [not safe](https://codeforces.com/blog/entry/61587). It may be acceptable in some cases, but in this particular case, it can severely backfire e.g. because on some systems it's always a 16-bit number.

Author: Was an interesting read, implemented the Mersenne twister

        int b = (rand()%points.size());
        s_next.points[a].swap(s_next.points[b]);
        return s_next;
    }

    double euclidean(pair<int,int> a, pair<int,int> b){

>         return pow(pow((a.first-b.first),2)+pow((a.second-b.second),2),0.5);

Reviewer: ```suggestion
        return std::hypot(a.first - b.first, a.second - b.second);
```

Reviewer: Please also put space between `)` and `{`, after `,`, and around arithmetic operators.

    }
    double E(){
        double dist = 0;
        bool first = true;
        int n = points.size();
        for(int i = 0;i < n; i++){
            dist += euclidean(points[i],points[(i+1)%n]);
        }
        return dist;
    };
};

int main(){
    pair<int,state> res;
    res = simAnneal();
    double E_best = res.first;
    state best = res.second;
    cout << "Lenght of shortest path found : " << E_best << "\n";
    cout << "Order of points in shortest path : \n";
    for(auto x: best.points){
        cout << x.first << " " << x.second << "\n";
    }
}
```

## Further modifications to the algorithm:

- Add a time based exit condition to the while loop to prevent TLE

> - The Probability acceptance function given above, prefers accepting states which are lower in energy because of the $E_{next} - E$ factor in the numerator of the exponent. You can simply remove this factor, to make the PAF independent of the difference in energies.

Reviewer: But why and when would one do it?

Author: Consider an E(s) which is pretty much constant for most points in its domain, but has very sharp drops for its minima's. In such a case, after finding a particular local minima, the PAF will mostly reject all points which are not inside another minima, as the energy difference is large. Thus the search will never exit a local minima and simulated annealing won't do much better than gradient descent. On the other hand, making the PAF independent of the difference in energies will allow the algorithm to do its magic.

- The effect of the difference in energies, $E_{next} - E$, on the PAF can be increased/decreased by increasing/decreasing the base of the exponent as shown below: 
```cpp
bool P(double E,double E_next,double T){

>     double e = 2; // set e to any real number greater than 1

Reviewer: Ahhh, an engineer wrote this, I see :grin:

    double prob = (double)rand()/RAND_MAX; // Generate a random number between 0 and 1
    if(pow(e,-(E_next-E)/T) > prob) return true;
    else return false;
}
```

## Problems

- [USACO Jan 2017 - Subsequence Reversal](https://usaco.org/index.php?page=viewproblem2&cpid=698)
- [Deltix Summer 2021 - DIY Tree](https://codeforces.com/contest/1556/problem/H)

> - [AtCoder Contest Scheduling](https://atcoder.jp/contests/intro-heuristics/tasks/intro_heuristics_a)

Reviewer: Did you test the implementation from the article on any of those?

Author: I've tested the implementation with the 3rd problem : [AtCoder Contest Scheduling](https://atcoder.jp/contests/intro-heuristics/tasks/intro_heuristics_a).

## Algorithm

We start with a random state $s$. In every step, we choose a neighbouring state $s_{next}$ of the current state $s$. If $E(s_{next}) < E(s)$, then we update $s = s_{next}$. Otherwise, we use a probability acceptance function $P(E(s),E(s_{next}),T)$ which decides whether we should move to $s_{next}$ or stay at $s$. T here is the temperature, which is initially set to a high value and decays slowly with every step. The higher the temperature, the more likely it is to move to $s_{next}$. 
At the same time we also keep a track of the best state $s_{best}$ across all iterations. Proceeding till convergence or time runs out.

<center>
<img src="https://upload.wikimedia.org/wikipedia/commons/d/d5/Hill_Climbing_with_Simulated_Annealing.gif" width="800px">
<br>
<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
<br>

> <i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.

Reviewer: ```suggestion
<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
</center>
```

## Algorithm

We start with a random state $s$. In every step, we choose a neighbouring state $s_{next}$ of the current state $s$. If $E(s_{next}) < E(s)$, then we update $s = s_{next}$. Otherwise, we use a probability acceptance function $P(E(s),E(s_{next}),T)$ which decides whether we should move to $s_{next}$ or stay at $s$. T here is the temperature, which is initially set to a high value and decays slowly with every step. The higher the temperature, the more likely it is to move to $s_{next}$. 
At the same time we also keep a track of the best state $s_{best}$ across all iterations. Proceeding till convergence or time runs out.

<center>
<img src="https://upload.wikimedia.org/wikipedia/commons/d/d5/Hill_Climbing_with_Simulated_Annealing.gif" width="800px">
<br>
<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
<br>

> <i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.

Reviewer: This misrenders in the [GitHub pages preview](https://gh.cp-algorithms.com/1371/num_methods/simulated_annealing.html), but renders correctly at the [built-in preview](https://cp-algorithms.com/preview.html). I'm not really sure what causes this, but, given that the image is under CC0, in principle, we can remove this line altogether.

</center>

> ### Temperature($T$) and decay($u$)

Reviewer: ```suggestion
### Temperature (T) and decay (u)
```
`$` misrenders in sidebar table of contents.

$T = u^{-N}$

### Example implementation for TSP
```cpp

class state {
    public:
    vector<pair<int, int>> points;
	std::mt19937 mt{ static_cast<std::mt19937::result_type>(
		std::chrono::steady_clock::now().time_since_epoch().count()
		) };
    state() {
        points = {%raw%} {{0,0},{2,2},{0,2},{2,0},{0,1},{1,2},{2,1},{1,0}} {%endraw%};
    }
    state next() {
        state s_next;
        s_next.points = points;
        uniform_int_distribution<> choose(0, points.size()-1);
        int a = choose(mt);
        int b = choose(mt);
        s_next.points[a].swap(s_next.points[b]);
        return s_next;
    }

    double euclidean(pair<int, int> a, pair<int, int> b) {
        return hypot(a.first - b.first, a.second - b.second);
    }
    
    double E() {
        double dist = 0;

>         bool first = true;

Reviewer: ```suggestion
```
It seems unused?

        int n = points.size();
        for (int i = 0;i < n; i++)
            dist += euclidean(points[i], points[(i+1)%n]);
        return dist;
    };
};

int main() {
    pair<double, state> res;
    res = simAnneal();
    double E_best = res.first;
    state best = res.second;
    cout << "Lenght of shortest path found : " << E_best << "\n";
    cout << "Order of points in shortest path : \n";
    for(auto x: best.points) {
        cout << x.first << " " << x.second << "\n";
    }
}
```

## Further modifications to the algorithm:

- Add a time based exit condition to the while loop to prevent TLE

> - The decay implemented above is an exponential decay. You can always replace this with a decay function as per your needs.

Reviewer: Is there any example where non-exponential decay performs better?

Author: Libraries which implement Sim Anneal usually provide options for non-exponential decay such as [this one](https://mlrose.readthedocs.io/en/stable/source/decay.html). A linear decay function will motivate the algorithm to explore actively even towards the end. This is useful when there are multiple local minima very close to each other.

- The Probability acceptance function given above, prefers accepting states which are lower in energy because of the $E_{next} - E$ factor in the numerator of the exponent. You can simply remove this factor, to make the PAF independent of the difference in energies.
- The effect of the difference in energies, $E_{next} - E$, on the PAF can be increased/decreased by increasing/decreasing the base of the exponent as shown below: 
```cpp
bool P(double E, double E_next, double T, mt19937 rng) {

>     e = 2 // set e to any real number greater than 1
>     double prob =  exp(-(E_next-E)/T);

Reviewer: ```suggestion
    double e = 2; // set e to any real number greater than 1
    double prob =  pow(e, -(E_next-E)/T);
```
